aspnetcore/performance/caching/middleware.md

```diff
@@ -1,15 +1,114 @@
+Expires | <p>A response deemed stale by this header will not be stored or retrieved unless overridden by other `Cache-Control` headers.</p>
+If-None-Match | <p>The full response will be served from cache if the value is not `*` and the `ETag` of the response doesn't match any of the values provided. Otherwise, a 304 (Not Modified) response will be served.</p>
+If-Modified-Since | <p>If the `If-None-Match` header is not present, a full response will be served from cache if the cached response date is newer than the value provided. Otherwise, a 304 (Not Modified) response will be served.</p>
+Date | <p>When serving from cache, the `Date` header is set by the middleware if it wasn't provided on the original response.</p>
+Content-Length | <p>When serving from cache, the `Content-Length` header is set by the middleware if it wasn't provided on the original response.</p>
+Age | <p>The `Age` header sent in the original response will be ignored. The middleware will compute a new value when serving a cached response.</p>
+
+## Troubleshooting
+If caching behavior is not as you expect, confirm that responses are cacheable and capable of being served from the cache by examining the request's incoming headers and the response's outgoing headers. The conditions by which a response will be cached or served from the cache are listed below.
+
+When testing and troubleshooting caching behavior, a browser may set request headers that affect caching in undesirable ways. For example, a browser may set the `Cache-Control` header to `no-cache` when you refresh the page. Instead of using a browser, use a tool like [Fiddler](http://www.telerik.com/fiddler), [Firebug](http://getfirebug.com/), or [Postman](https://www.getpostman.com/), all of which allow you to explicitly set request headers.
+
+* The request must result in a 200 (OK) response from the server.
+* The request method must be GET or HEAD.
+* Terminal middlewares, such as Static File Middleware, must not process the response prior to the Response Caching Middleware.
+* The Authorization header must not be present.
+* `Cache-Control` header parameters must be valid, and the response must be marked `public`.
+* The `Pragma: no-cache` header/value must not be present if the `Cache-Control` header is not present, as the `Cache-Control` header overrides the `Pragma` header when present.
+* The `Set-Cookie` header must not be present.
+* `Vary` header parameters must be valid and not equal to `*`.
+* The `Content-Length` header value (if set) must match the size of the response body.
+* The `If-None-Match` header value (if set) must not be `*`, and the `ETag` of the response must not match any of the values provided by the header.
```

Comment:
We don't check the If-None-Match and If-Modified-Since headers on storage, those are checked on retrieval in subsequent requests. However, we do check and make sure the entry is fresh upon storage against headers like `Cache-Control: max-age=<value>` or `Expires: <value>`.